plugins/radar-android-google-sleep/src/main/java/org.radarbase.passive.google.sleep/GoogleSleepManager.kt

```diff
@@ -0,0 +1,156 @@
+    private val sleepHandler = SafeHandler.getInstance("Google Sleep", Process.THREAD_PRIORITY_BACKGROUND)
+    private val sleepPendingIntent: PendingIntent
+    private val isPermissionGranted
+    get() = ContextCompat.checkSelfPermission(service,ACTIVITY_RECOGNITION_COMPAT) == PackageManager.PERMISSION_GRANTED
+
+    init {
+        name = context.getString(R.string.googleSleepDisplayName)
+        sleepBroadcastReceiver = SleepReceiver(this)
+        sleepPendingIntent = createSleepPendingIntent()
+    }
+
+    override fun start(acceptableIds: Set<String>) {
+        register()
+        sleepHandler.start()
+        status = SourceStatusListener.Status.READY
+        sleepHandler.execute {
+            registerForSleepData()
+        }
+    }
+
+     fun sendSleepClassifyData(sleepIntent: Intent) {
+        logger.info("Sleep segment event data received")
+        val sleepSegmentEvents: List<SleepSegmentEvent> = SleepSegmentEvent.extractEvents(sleepIntent)
+        sleepSegmentEvents.forEach { sleepSegmentEvent ->
+            val time = currentTime
+            val sleepStartTime: Double = sleepSegmentEvent.startTimeMillis / 1000.0
+            val sleepEndTime: Double = sleepSegmentEvent.endTimeMillis / 1000.0
+            val sleepDuration: Double = sleepSegmentEvent.segmentDurationMillis / 1000.0
+            val sleepStatus: SleepStatus = sleepSegmentEvent.status.toSleepStatus()
+
+            send(segmentEventTopic, GoogleSleepSegmentEvent(sleepStartTime, time, sleepStartTime,
+            sleepEndTime, sleepDuration, sleepStatus))
+        }
+    }
+
+     fun sendSleepSegmentData(sleepIntent: Intent) {
+        logger.info("Sleep classify event data received")
+        val sleepClassifyEvents: List<SleepClassifyEvent> = SleepClassifyEvent.extractEvents(sleepIntent)
+        sleepClassifyEvents.forEach {  sleepClassifyEvent ->
+            val time = sleepClassifyEvent.timestampMillis / 1000.0
+            val sleepConfidence = sleepClassifyEvent.confidence
+            val light = sleepClassifyEvent.light
+            val motion = sleepClassifyEvent.motion
+
+            send(classifyEventTopic, GoogleSleepClassifyEvent(time, currentTime, sleepConfidence, light, motion))
+        }
+    }
+
+    @SuppressLint("MissingPermission")
+    private fun registerForSleepData() {
+        if (isPermissionGranted) {
+            status = SourceStatusListener.Status.CONNECTING
+
+            ActivityRecognition.getClient(service)
+                .requestSleepSegmentUpdates(
+                    sleepPendingIntent,
+                    SleepSegmentRequest.getDefaultSleepSegmentRequest() )
+                .addOnSuccessListener {
+                    status = SourceStatusListener.Status.CONNECTED
+                    logger.info("Successfully subscribed to sleep data")
+                }
+                .addOnFailureListener {exception ->
+                    status = SourceStatusListener.Status.DISCONNECTED
```

Comment:
Please use `disconnect()` instead.
```suggestion
                    disconnect()
```
Directly setting the status to `DISCONNECTED` will mess with the logic in `SourceService`.